curl: default to darwinssl to avoid OpenSSL licence tainting dependencies

danchr opened the pull request:
The licence taint prevents the builders from uploading binary archives of any GPL package that depends on cURL. This affects at least the Git and QEMU ports, and probably others too. See for example:

https://build.macports.org/builders/ports-10.15_x86_64-builder/builds/31774/steps/gather-archives/logs/stdio

This is a rather invasive change, and it might not be desirable, but I figured I'd post it here as an alternative to #7484.

#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk -v ORS=' ' '{print $NF}')"
-->
macOS 10.x
Xcode 8.x

###### Verification <!-- (delete not applicable items) -->
Have you

- [ ] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [ ] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [ ] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
<!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [ ] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

ryandesign commented on net/curl/Portfile:
```diff
@@ -34,7 +34,7 @@ checksums                       ${curl_distfile} \
     PortGroup                   muniversal 1.0
 
-    revision                    0
+    revision                    1
```
The revision should not be increase because this doesn't change the files that will get installed for anyone who already has it installed.

danchr replied:
Is there another way to trigger a rebuild, or do the bots just do that when you change the file?

ryandesign replied:
The buildbot will build and upload anything that has not already been built and uploaded. When I wrote that comment, you were proposing changing the default variant to +darwinssl. The set of variants is part of the archive file name so in this case the buildbot would check if an archive for this version and revision of curl with +darwinssl had already been uploaded, would find that it had not been, and would build and upload it.

Since then, you've changed the PR to propose that the +ssl variant be removed. +ssl was formerly a default variant so if you use this new proposal then yes the revision should be increased.

ryandesign replied:
In fact even if it had not been a default variant, removing a variant always requires increasing the revision, otherwise there could still be users out there with that variant installed.

ryandesign commented on net/curl/Portfile:
```diff
@@ -192,10 +192,23 @@ if {${name} eq ${subport}} {
+            notes "
+Please note that the default SSL backend was recently changed from OpenSSL to \
+the native Secure Transport backend. To reenable binary packages, run:
+  $ ${0} upgrade --enforce-variants curl -ssl +darwinssl
```
This doesn't work: `can't read "0": no such variable`. In other ports, we just write it out as `sudo port`. I don't think there's a better way right now.

ryandesign commented on net/curl/Portfile:
```diff
@@ -192,10 +192,23 @@ if {${name} eq ${subport}} {
+        # remove after 2021-07-01
+        if {${os.platform} eq "darwin"} {
+            notes "
+Please note that the default SSL backend was recently changed from OpenSSL to \
```
In notes, don't put a space before the backslash at the end of the line because that will turn into a double space when printed on the terminal.

danchr commented:
I've pushed an updated version of the patch — rather than adding a note, I removed the `darwinssl` variant and renamed `ssl` to `openssl`. That should ensure a smooth upgrade path, shouldn't it?

Here's the full commit message:

> curl: fold darwinssl variant into port; use it on Mavericks and later
> 
> This avoids numerous licensing conflicts in GPL ports that depend on
> cURL, such as Git and QEMU ports.
> 
> The Secure Transport is the native backend, and cURL without TLS isn't
> all that useful, so I simply folded into the port. We want to ensure a
> smooth upgrade experience, so I renamed the `ssl` variant to
> `openssl`. This means that anyone on a recent macOS will be switched
> to native TLS, and everyone else just sticks to the default.
> 
> A few ports explicitly checked for the `ssl` variant of cURL. Now that
> all variants actually provide SSL/TLS,[1] I doubt they're needed.
> 
> [1] Except on FreeBSD, but I'm not sure that actually matters. Should
>     anyone actually be using MacPorts there, I'm sure they can manage
>     without the extra error-checking.

danchr commented:
> Thanks, I was thinking about making such a change awhile ago, for the same reason. I didn't do it yet because I hadn't yet tested if it is compatible with all macOS versions back to 10.4. Assuming it is not, we should only default to darwinssl on macOS versions where that will work. Alternately, we could default to darwinssl only on macOS versions where Apple's curl uses darwinssl (Mavericks and later).

I remember running into some issues when using the `+darwinssl` variant some time ago, but haven't seen it lately. The issue might have been valid, where Safari wasn't able to connect to the server either, but I don't remember the exact details.

One option is to change the default on older systems to the `gnutls` variant so that they also binary archives for Git, etc. Renaming the `ssl` variant should allow us to do that as well.

ryandesign commented:
> I've pushed an updated version of the patch — rather than adding a note, I removed the `darwinssl` variant and renamed `ssl` to `openssl`. That should ensure a smooth upgrade path, shouldn't it?

I'm not a fan of removing the darwinssl variant and making it always-on. That removes the ability for the user to turn ssl off, which is currently possible, and it made you repeat the code `configure.args-delete --without-ca-bundle --with-darwinssl` in the other variants, and repeating code is bad.

It's also a matter of UI. When I use variants to provide the user with a set of options from which they may choose one, I like to have a variant for every choice and not leave one out. I think of this usage of variants like radio buttons in a GUI, and I would rather have this set of options:

```
(*) Use darwin SSL
( ) Use gnutls
( ) Use openssl
( ) Use wolfssl
```

than this set:

```
( ) Use gnutls instead of darwin SSL
( ) Use openssl instead of darwin SSL
( ) Use wolfssl instead of darwin SSL
```

In short, I'd favor keeping the darwinssl variant.

Renaming the ssl variant to openssl would be seen by MacPorts as removing the ssl variant and adding a new openssl variant, and would automatically upgrade all current users of the now-removed ssl variant to the new defaults, be they darwinssl, gnutls or something else, which was your intention and seems like it could be fine. Of course it would require all users who actually want specifically openssl support to reenable it using the openssl variant. I would guess there are few such users so that's probably fine, assuming there are no ports out there that actually require curl with openssl. Do we know whether or not that is the case? For example, you removed the code from centerim, py-isce2, and shibboleth that enforced the ssl variant. Did you verify that these ports are compatible with non-openssl ssl versions of curl?

> I remember running into some issues when using the `+darwinssl` variant some time ago, but haven't seen it lately. The issue might have been valid, where Safari wasn't able to connect to the server either, but I don't remember the exact details.

I tried installing curl +darwinssl on 10.6 and 10.7. It does compile there and works with compatible servers. But I see what you mean: trying to access e.g. https://github.com or https://sourceforge.net doesn't work (`curl: (35) A bad protocol version was encountered`), whereas it does when using the openssl variant. I also tested on 10.8 and 10.9 and darwinssl works fine there with those servers. This is probably the same problem that we have when using /usr/bin/curl on OS X 10.8 and earlier: the openssl that /usr/bin/curl uses on 10.8 and earlier, and apparently the security framework on 10.7 and earlier, doesn't support the modern security protocols [now required by github](https://github.blog/2018-02-23-weak-cryptographic-standards-removed/) and sourceforge and probably other sites. So we should only default to darwinssl on 10.8 and later.

kencu commented:
OpenSSL 3.x is coming out soon, and has a new Apache2 license. 

I believe that will make all these openssl license headaches go away. 

If it turns out that is all correct and will work like I think, perhaps we should scope out the timeline and ride it out? We've been living with this situation for 15 years. Be a shame to change everything now, and then change it all back again to the way it is right now.

ryandesign commented:
They've been talking about the new apache-2-licensed openssl for years. Do we really think it's coming out soon?

danchr commented:
> Do we really think it's coming out soon?

The current plan seems to be [2020 early Q4](https://wiki.openssl.org/index.php/OpenSSL_3.0_Release_Schedule). Since they’re using quality-based criteria, that might well slip. It’s worth noting that `http2` support — another desirable default — also depends on OpenSSL.

Personally, I think we should change the default regardless of licence; using Secure Transport is consistent with the OS distribution, and makes it consistent with the built-in and custom roots configured in the Keychain.

kencu commented:
if you do change it, +darwinssl is a dying beast on any system not actively updated by Apple, so openssl should be the default for -- what -- 10.12 and less? Or is it 10.11 right now.

ryandesign commented:
Hm. So if we default to darwinssl on Apple-supported OS versions, you propose changing that default to openssl as soon as Apple drops support. If we do that, and if we do as I suggested above and keep the darwinssl variant, that will give a suboptimal experience for upgrading users unless we implement https://trac.macports.org/ticket/46956. If we don't implement that, the better user experience would be to keep the defaults the same for all OS versions, or to remove the darwinssl variant and just make that the default behavior of the port when no other ssl variant is selected, as already proposed in the PR.

danchr commented:
> It’s worth noting that `http2` support — another desirable default — also depends on OpenSSL.

I created #7641 to address that, although it has some conflicts on upgrade, unfortunately.

> If we do that, and if we do as I suggested above and keep the darwinssl variant, that will give a suboptimal experience for upgrading users unless we implement https://trac.macports.org/ticket/46956.

Yeah, that's the main motivation for removing it; it allows us to relatively easily change the defaults back and forth. Alternatively, we could perform a bit of magic where we use different names for the variant based on whether `+darwinssl` is available. I'm not sure that's an improvement, though…

danchr commented:
I've added a commit that reverts most of the changes. Combined `+ssl` is simply renamed to `+openssl`, and `+darwinssl` is removed from older systems. (I still think it's a bit silly to have a variant that adds no dependencies, but I'm not the maintainer of the port 🙂)

If this is an acceptable compromise, I'll fold it into a single commit and check the dependents.

danchr commented:
Ping maintainers of other ports: @nerdling @scantor @piyushrpt @mazhe

scantor commented:
FWIW, my dependent port requires the OpenSSL variant. Is the change here just to start using "openssl" as the curl variant name? Is that clear to change now, or is this waiting on further decisions?

nerdling commented:
@ryandesign anything you need from us other maintainers?

danchr commented:
Closing; the change is too invasive.
